golang/cosmos/x/swingset/types/store.go

```diff
@@ -0,0 +1,155 @@
+package types
+
+import (
+	"bytes"
+	"io"
+
+	prefixstore "github.com/cosmos/cosmos-sdk/store/prefix"
+	storetypes "github.com/cosmos/cosmos-sdk/store/types"
+)
+
+var (
+	// SwingStoreValueSuffix is the null byte appended to all SwingStore values
+	// before writing them in the underlying KVStore. This allows accepting nil
+	// values, which are expected of the JS swing-store but disallowed by a
+	// normal KVStore.
+	SwingStoreValueSuffix = []byte{0}
+)
+
+var _ storetypes.KVStore = SwingStore{}
+
+// SwingStore wraps a KVStore, possibly prefixed, allowing nil values.
+// It does so by appending a null byte to all values before writing them in the
+// parent store, and stripping a trailing byte when reading values from the
+// parent store.
+type SwingStore struct {
+	parent storetypes.KVStore
+}
+
+// NewSwingStore creates a SwingStore from a given KVStore and an optional prefix
+func NewSwingStore(parent storetypes.KVStore, prefix []byte) SwingStore {
+	if prefix != nil {
+		parent = prefixstore.NewStore(parent, prefix)
+	}
+	return SwingStore{parent}
+}
```

Comment:
I think it would be a slightly better design to handle key prefixing one level up such that SwingStore handles only expansion of the value domain rather than that _and_ prefixing (which is already covered by `cosmos/cosmos-sdk/store/prefix`).
```suggestion
// NewSwingStore creates a SwingStore from a given KVStore.
func NewSwingStore(kvStore storetypes.KVStore) SwingStore {
	return SwingStore{kvStore}
}
```
```diff
-	store := ctx.KVStore(k.storeKey)
-	return types.NewSwingStore(store, []byte(swingStoreKeyPrefix))
+	parentStore := ctx.KVStore(k.storeKey)
+	store := prefixstore.NewStore(parentStore, []byte(swingStoreKeyPrefix))
+	return types.NewSwingStore(store)
```

It might then also rename SwingStore to something more general like NillableValueStore (if it survives at all, see other comments for reasons why underlying Store support for empty `[]byte{}` values might make it unnecessary).